src/ceph-volume/ceph_volume/activate/main.py

```diff
@@ -0,0 +1,87 @@
+# -*- coding: utf-8 -*-
+
+import argparse
+
+from ceph_volume import terminal
+from ceph_volume.devices.lvm.activate import Activate as LVMActivate
+from ceph_volume.devices.raw.activate import Activate as RAWActivate
+from ceph_volume.devices.simple.activate import Activate as SimpleActivate
+
+
+class Activate(object):
+
+    help = "Activate an OSD"
+
+    def __init__(self, argv):
+        self.argv = argv
+
+    def main(self):
+        parser = argparse.ArgumentParser(
+            prog='ceph-volume activate',
+            formatter_class=argparse.RawDescriptionHelpFormatter,
+            description=self.help,
+        )
+        parser.add_argument(
+            '--osd-id',
+            help='OSD ID to activate'
+        )
+        parser.add_argument(
+            '--osd-uuid',
+            help='OSD UUID to active'
+        )
+        parser.add_argument(
+            '--no-systemd',
+            dest='no_systemd',
+            action='store_true',
+            help='Skip creating and enabling systemd units and starting OSD services'
+        )
+        parser.add_argument(
+            '--no-tmpfs',
+            action='store_true',
+            help='Do not use a tmpfs mount for OSD data dir'
+        )
+        self.args = parser.parse_args(self.argv)
+
+        # first try raw
+        try:
+            RAWActivate([]).activate(
+                device=None,
+                start_osd_id=self.args.osd_id,
+                start_osd_uuid=self.args.osd_uuid,
+                tmpfs=not self.args.no_tmpfs,
+                systemd=not self.args.no_systemd,
+                block_wal=None,
+                block_db=None,
+            )
+            return
+        except Exception as e:
+            terminal.info(f'Failed to activate via raw: {e}')
+
+        # then try lvm
+        try:
+            LVMActivate([]).activate(
+                argparse.Namespace(
+                    osd_id=self.args.osd_id,
+                    osd_fsid=self.args.osd_uuid,
+                    no_tmpfs=self.args.no_tmpfs,
+                    no_systemd=self.args.no_systemd,
+                )
+            )
+            return
+        except Exception as e:
+            terminal.info(f'Failed to activate via lvm: {e}')
```

Comment:
@liewegas it turns out that if we are dealing with bluestore osds, `RAWActivate().activate()` is able to activate them.
it means it would enter this block when either activating filestore osds (given raw doesn't support filestore) or   bluestore dmcrypt osds if the dm is closed (the prepare operation leaves the dm opened, but after a reboot for instance?).

Reply:
right.  is this a problem?  IIUC it doesn't matter whether raw or lvm does the activating, as long as it gets activated...

Reply:
no, this isn't a huge problem, I personally think this is not very elegant.
Regardless, it can be a bit confusing, especially when doing some troubleshoot/debug I think.

Reply:
We could potentially write a global refactor regarding osd activation but that would be a lot of changes. I'm unsure this is worth it at the moment.